shelljs/shelljs.d.ts

```diff
@@ -434,31 +436,44 @@ declare module "shelljs"
      * Object containing environment variables (both getter and setter). Shortcut to process.env.
      */
     export var env: { [key: string]: string };
-
-    /*
-
-    // Not yet implemented due to implementation issues (constant overloads and return types).
-    // See: https://github.com/arturadib/shelljs#execcommand--options--callback
-
-    export function exec(command: string, options: ExecOptions, callback: (code: number, output: string) => any): any;
-    export function exec(command: string, options: ExecOptions): any;
-
-    interface ExecOptions
-    {
-        silent: boolean;
-        async: boolean;
-    }
-
-    */
-
+    
     /**
      * Executes the given command synchronously.
-     * @param  {string}          command The commadn to execute.
+     * @param  {string}          command The command to execute.
      * @return {ExecReturnValue}         Returns an object containing the return code and output as string.
      */
-    export function exec(command: string): ExecReturnValue;
+    export function exec(command: string): ExecOutputReturnValue;
+    /**
+     * Executes the given command synchronously.
+     * @param  {string}          command The command to execute.
+     * @param  {ExecOptions}     options Silence and synchronous options.
+     * @return {ExecReturnValue}         Returns an object containing the return code and output as string.
+     */
+    export function exec(command: string, options: ExecOptions): ExecOutputReturnValue;
```

Comment:
This still isn't quite correct. The return type of this method can be `ExecOutputReturnValue` or `ChildProcess`. The latter case isn't covered by that definition.

There might be an ["overload-on-constant"](http://blogs.msdn.com/b/typescript/archive/2013/03/25/working-on-typescript-0-9-generics-overload-on-constants-and-compiler-performance.aspx) approach, but I've not (yet?) figured out whether it is actually possible to use an interface with overloads-on-constants.
It may look somehow like this:
```TypeScript
// if async == true -> ChildProcess
export function exec(command: string, options: { async: true; silent?: boolean}): child.ChildProcess;
// if async == false -> ExecOutputReturnValue
export function exec(command: string, options: { async: false; silent?: boolean}): ExecOutputReturnValue;

export function exec(command: string, options: ExecOptions): ???; // there must be a common type (which matches ChildProcess and ExecOutputReturnValue)
```

If this isn't possible, we should at least state that both types can be returned (and let the caller filter/cast):
```TypeScript
export function exec(command: string, options: ExecOptions): ExecOutputReturnValue|child.ChildProcess;
```
This was actually the issue why I haven't implemented this function more strongly typed than "any".